### PR TITLE
Fixes juju kubernetes master: 1. Get certs from a dead leader. 2. Append tokens.

### LINKED ISSUE #43519: Remove and re-add juju kubernetes-master units breaks the cluster
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Kubernetes version** (use `kubectl version`): 1.5

**Environment**: Any juju deployed environment
- **Cloud provider or hardware configuration**: Any cloud used by juju
- **OS** (e.g. from /etc/os-release): Ubuntu 16.04.2 LTS
- **Kernel** (e.g. `uname -a`): 4.4.0-66-generic
- **Install tools**:
- **Others**:

**What happened**:
Removed all kubernetes-master units and then re-added one master node and that master was failing.

**What you expected to happen**:
The master would function properly.

**How to reproduce it** (as minimally and precisely as possible):
Remove all kubernetes-master units and then re-added one master. There is also this tests that shows the issue: https://github.com/ktsakalozos/bundle-canonical-kubernetes/blob/bug/master-check/tests/40-remove-add-master-check.py

**Anything else we need to know**:
This is for the deployments done with juju

### LINKED ISSUE #43563: known_tokens.csv only has the latest entry pushed in the juju kubernetes-master charm
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT 

**Kubernetes version** (use `kubectl version`): 1.5

**Environment**:
- **Cloud provider or hardware configuration**: Any cloud used do deploy kubernetes with Juju
- **OS** (e.g. from /etc/os-release): Ubuntu 16.04
- **Kernel** (e.g. `uname -a`): 4.4.0-66-generic
- **Install tools**: Juju
- **Others**:

**What happened**:
/srv/kubernetes/known_tokens.csv contains only one token

**What you expected to happen**:
/srv/kubernetes/known_tokens.csv to have all tokens we append here: https://github.com/kubernetes/kubernetes/blob/master/cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py#L164

**How to reproduce it** (as minimally and precisely as possible):
Deploy a kubernetes master using Juju

**Anything else we need to know**:
Here we assume we append tokens:
https://github.com/kubernetes/kubernetes/blob/master/cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py#L164
But we truncate the file when we open it: https://github.com/kubernetes/kubernetes/blob/master/cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py#L763

### PR DESCRIPTION
**Pull request by ktsakalozos**
**What this PR does / why we need it**:
Fixes two issues with the Juju kubernetes master.

1. Grab certificates from a leader that is already removed.
2. Append (not truncate) auth tokens 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
fixes #43563 fixes #43519

**Special notes for your reviewer**:

**Release note**:

```
Recover certificates from leadership context in case all masters die in a Juju deployment
```

### CODE REVIEW

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 1]
```diff
@@ -199,15 +209,27 @@ def setup_non_leader_authentication():
...
+    api_opts.add('--basic-auth-file', basic_auth)
+    api_opts.add('--token-auth-file', known_tokens)
+    api_opts.add('--service-cluster-ip-range', service_cidr())
+    api_opts.add('--service-account-key-file', service_key)
+    controller_opts.add('--service-account-private-key-file', service_key)
+
+    set_state('authentication.setup')
+
+
+def get_keys_from_leader(keys):
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: No docstring for this method

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 2]
```diff
@@ -199,15 +209,27 @@ def setup_non_leader_authentication():
...
     os.makedirs('/etc/kubernetes', exist_ok=True)
     os.makedirs('/srv/kubernetes', exist_ok=True)
 
-    hookenv.status_set('maintenance', 'Rendering authentication templates.')
-
-    # Set an array for looping logic
-    keys = [service_key, basic_auth, known_tokens]
     for k in keys:
         # If the path does not exist, assume we need it
         if not os.path.exists(k):
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: If the path doesn't exist and the contents are none this new method now returns, which is different behavior than a return in the original method. Is this intended?

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py (hunk 1 above)
Comment (ktsakalozos) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Done!

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py (hunk 2 above)
Comment (ktsakalozos) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: You are right. In the latest commit get_keys_from_leader returns False if any of the keys were not fetched

**Comment by mbruzek** [comment]
Thanks for changing the code @ktsakalozos!

/approve

**Comment by marcoceppi** [comment]
@k8s-bot ok to test

**Comment by marcoceppi** [comment]
Can you please add release notes for this? Thanks.

**Comment by ktsakalozos** [comment]
I've addressed conflicts and updated the release note. Thanks

**Pull request merged**